docs/fpc-key-dist.puml

```diff
@@ -1,97 +1,319 @@
+  note left of TLCC_Enclave1
+    The actual implementation of check_committed_data
+    is out of scope here. A possible solution can be
+    sketched as follows. A principal asks TLCC whether
+    a chaincode (say ERCC) has stored a key
+    (say ECC2_SPK||".registration"), with a specific hash
+    value (say h=H(m || attestation_report)). TLCC returns
+    true/false (say h==H(value of store key)).
   end note
-
-  Enclave1 -> Enclave1 : SIG_ECC <- sign {{m, CC_PK}} with ECC_S_SK
-  note right
-    enclave signs the encrypted message and the chaincode pubkey.
-    - SIG_ECC denotes the signature
-
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> TLCC_Enclave1    ++: can_endorse(ECC2_S_PK)
+    note right
+        TLCC checks the endorsement policy to determine
+        if the recipient enclave is an eligible endorser.
+    end note
+    note right
+        TLCC should check that the recipient enclave's MRENCLAVE
+        matches the chaincode definition's MRENCLAVE.
+        This could enable key distribution from an old cc version,
+        with a different MRENCLAVE.
+        Notice: in this case there cannot be also a check
+        can_endorse(ECC1_S_PK) -- or it should be negated.
+    end note
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> Enclave1         :   check CC_name, CC_ver, CH_id, org_id in sealed data
+  Enclave1      -> Enclave1         :   check ECC2_S_PK, ECC_E_PK, attestation_report
+  Enclave1      -> Enclave1         :   m <- encrypt <SEK, CC_E_SK> with ECC2_E_PK
+    note left
+        enclave encrypts state encryption key and chaincode private key
+        - m denotes encrypted message
+    end note
+
+  Enclave1      -> Enclave1         :   SIG_ECC1 <- sign <ECC1_S_PK, ECC2_s_PK, CC_E_PK, m> with ECC1_S_SK
+    note left
+        enclave signs the key material
+        it distributes together with the
+        sender and intended recipient identity
+    end note
+
+  Enclave1      -> Enclave1         :   w <- <ECC1_S_PK, ECC2_s_PK, CC_E_PK, SIG_ECC1>
+
+  return m, w /'ECC1 -> Enclave1'/
+  return m, w /'ECC_Manager1 -> ECC1'/
+  return m, w /'Admin1 -> ECC_Manager1'/
+
+  note right Admin1
+    Key distribution is in progress. We need a means to transfer the message with
+    the encrypted keys to the recipient enclave. The next step uses the ledger
+    itself for this data transfer. This additionally allows to log the two phases
+    of the key distribution: (1) the sender enclave produced the encrypted data blob;
+    (2) the recipient enclave correctly received it. This can be useful when dealing
+    with enclaves of two different orgs.
   end note
 
-  return ECC_S_PK, SIG_ECC, m, CC_PK
-  return ECC_S_PK, SIG_ECC, m, CC_PK
-  return ECC_S_PK, SIG_ECC, m, CC_PK
+  group start key distribution endorsement
```

Comment:
publication of CCKeys sharing

Reply:
that's an invocation for an actual endorsement in fabric, which is repeated with other ERCCs (see box below) before sending a transaction to the orderer